### DISCUSSION TITLE
How to use serialized spancontext as parent, so that link can be established between two spans

### DISCUSSION THREAD

**Discussion by MacNale** [open-telemetry/opentelemetry-java#3023] [Q&A] [answered]
I want to recreate a spancontext serialized into a file. e.g
```
private static SpanContext createContext() {
    SpanContext remoteSpanContext = SpanContext.createFromRemoteParent("12345678876543211234567887654321",
        "8765432112345678", TraceFlags.getSampled(), TraceState.getDefault());

    return remoteSpanContext;
}
```

Once the SpanContext is created, I wish to use this as if it came from REMOTE service to create subsequent spans.
```
addLegacyServiceSpans(createContext());

private static void addLegacyServiceSpans(SpanContext spanContext) {
  OpenTelemetry openTelemetryParent = ExampleConfiguration.initOpenTelemetry("MYSERVICE");
  Tracer trace = openTelemetryParent.getTracer("MYTRACE");
  Span parentSpan = trace.spanBuilder("MYNEWOPERATION")
    .setStartTimestamp(Instant.now().toEpochMilli(), TimeUnit.MILLISECONDS)
    .startSpan();
}
```

I am not sure, how 'parentSpan' can use the passed 'spanContext' as remote span context.

**Comment by jkwatson** [comment, marked as answer]
`context.with(Span.wrap(spanContext))` will get you a `Context` instance that can be used as the argument to the `setParent(Context)` call on the SpanBuilder.

**Reply by surajeetdev** [reply]
But the Span created from Span.wrap is no-op and any span created using that is alo no-op. How to create a Span using a SpanContext which is NOT non-op?

**Reply by jkwatson** [reply]
The results of `Span.wrap()` can be used to populate a context (as above) which can be used as the parent of a newly created local span. You can't "continue" a span from an upstream system in OpenTelemetry, if that's what you're looking to do.

**Reply by surajeetdev** [reply]
Thats what i was trying to do , excatly what you suggested , not "continue" a span from upstream

However the child span also becomes no-op . Nothing is exported and when i print out childSpan.isRecording() , it prints false. 

Code snippet:

SpanContext spanContext = SpanContext.create(traceId,spanId,TracFlags.getDefault(),TraceState.getDefault())
Span parentSpan = Span.wrap(spanContext)
Span childSpan = tracer.spanBuilder(String.valueOf("execute")).setParent(Context.current().with(parentSpan)).startSpan();

**Reply by jkwatson** [reply]
TraceFlags.getDefault() has sampled = false in it, which, depending on your Sampler, will probably drop any subsequent spans

**Reply by surajeetdev** [reply]
ahh that was the reason , will change it to TraceFlags.getSampled() and try. 
Thanks

**Reply by surajeetdev** [reply]
I created SpanContext using "Span.createFromRemoteParent" using the traceid and spanid received from upstream , and then a span from that SpanContext . 

However it didn't create a child span rather a top level span. Am i missing something?

**Reply by jkwatson** [reply]
I'd need to see some code to know if you're missing something. :)

**Reply by surajeetdev** [reply]
the traceIdhex and spanIdHex below is from upstream. 


                        SpanContext spanContext = SpanContext.createFromRemoteParent(traceIdHex, spanIdHex, TraceFlags.getSampled(), TraceState.getDefault());
                        
                        logger.info("CreatedSpanContext with traceId {} spanId {} ",spanContext.getTraceId(),spanContext.getSpanId());
                        Span spanNoOp =Span.wrap(spanContext);
                        
                        OpenTelemetry telemetry = GlobalOpenTelemetry.get();
                        Tracer tracer= telemetry.getTracerProvider().get("tracifer");
                        
                        Span span = tracer.spanBuilder(String.valueOf("flatbuffer_deserializer")).setParent(Context.current().with(spanNoOp)).startSpan();                        
                        try (Scope scope = span.makeCurrent()){
                         
                        }finally {
                            span.end();
                        }

**Reply by jkwatson** [reply]
I did this locally and it seems to have worked as desired. You're not seeing the trace id of the span being the same as the trace id that you pass in on the spanContext?

Here is my code :
```
    String traceIdHex = TraceId.fromLongs(23434, 343434);
    String spanIdHex = SpanId.fromLong(2343434);
    SpanContext spanContext = SpanContext.createFromRemoteParent(traceIdHex, spanIdHex, TraceFlags.getSampled(), TraceState.getDefault());

    System.out.println("CreatedSpanContext with traceId " + spanContext.getTraceId() + " spanId " + spanContext.getSpanId());
    Span spanNoOp =Span.wrap(spanContext);

    Span span = tracer.spanBuilder("flatbuffer_deserializer").setParent(Context.current().with(spanNoOp)).startSpan();
    try (Scope scope = span.makeCurrent()){

    }finally {
      span.end();
    }
```

And the output (with a logging span exporter installed in the SDK):

```
CreatedSpanContext with traceId 0000000000005b8a0000000000053d8a spanId 000000000023c20a
INFO: 'flatbuffer_deserializer' : 0000000000005b8a0000000000053d8a 5bbdac2a4690ba28 INTERNAL [tracer: ConfigureTraceExample:] {}
```

**Reply by surajeetdev** [reply]
Actually i do se the traceid and spanId  in the logs (printed from the SpanContext that was created) to be the same i used to create the SpanContext.

I guess the SpanExporter might be messing up , as i don't see it as child of remote in the Jaeger UI

**Reply by jkwatson** [reply]
The remote wrapped span you created will never by exported by this process. It is assumed that an upstream process has already exported it somewhere.

**Reply by ukrcoder** [reply]
Tried the same with quarkus 3.4.3 and latest opentelemetry and it didn't add the parent. Any info maybe anything is changed? The problem is I don't have header traceparent but I'm able to get it in cookies and invoke this code snippet inside filter.
